com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/JarSignatureVerifier.java

```diff
@@ -0,0 +1,175 @@
+                                                       getTrustedCertificatesNames(targetCertificates)));
+        }
+    }
+
+    private List<X509Certificate> toX509Certificates(Certificate[] certificates) {
+        return Arrays.stream(certificates)
+                     .map(certificate -> (X509Certificate) certificate)
+                     .collect(Collectors.toList());
+    }
+
+    private boolean isValidCertificateChain(List<X509Certificate> targetCertificates, List<X509Certificate> candidateCertificateChain,
+                                            String certificateCN) {
+        boolean certificateCNMatches = false;
+        boolean targetCertificatesMatchEntryCertificate = false;
+        for (X509Certificate certificate : candidateCertificateChain) {
+            checkValidityOfCertificate(certificate);
+            if (!certificateCNMatches) {
+                certificateCNMatches = certificateCNMatches(certificate, certificateCN);
+            }
+            if (!targetCertificatesMatchEntryCertificate) {
+                targetCertificatesMatchEntryCertificate = doTargetCertificatesMatchEntryCertificate(targetCertificates, certificate);
+            }
+        }
+        if (!certificateCNMatches) {
+            throw new SLException(MessageFormat.format(Messages.NONE_OF_YOUR_CERTIFICATES_CNS_MATCH, certificateCN));
+        }
+        return targetCertificatesMatchEntryCertificate;
+    }
+
+    private void checkValidityOfCertificate(X509Certificate x509Certificate) {
+        try {
+            x509Certificate.checkValidity();
+        } catch (CertificateExpiredException | CertificateNotYetValidException e) {
+            throw new SLException(e, e.getMessage());
+        }
+    }
+
+    private boolean doTargetCertificatesMatchEntryCertificate(List<X509Certificate> targetCertificates, X509Certificate certificate) {
+        return targetCertificates.stream()
+                                 .anyMatch(currentCertificate -> Objects.equals(currentCertificate, certificate));
+    }
+
+    private boolean certificateCNMatches(X509Certificate certificate, String certificateCN) {
+        if (certificateCN == null) {
+            return true;
+        }
+        Map<String, String> params = getCertificateSubjectParameters(certificate);
+        return Objects.equals(params.get(CN), certificateCN);
+    }
+
+    private Map<String, String> getCertificateSubjectParameters(X509Certificate certificate) {
+        Pattern subjectSeparatorPattern = Pattern.compile(", *");
+        return Splitter.on(subjectSeparatorPattern)
+                       .withKeyValueSeparator("=")
+                       .split(certificate.getSubjectDN()
+                                         .getName());
+    }
+
+    private List<String> getTrustedCertificatesNames(List<X509Certificate> targetCertificates) {
+        return targetCertificates.stream()
+                                 .map(this::getCertificateSubjectParameters)
```

Comment:
This can be extracted in a separate method `getCertificateCN`. You're using it in 2 places.

Reply:
I meant to have the following method:
```
    private String getCertificateCN(X509Certificate certificate) {
        Map<String, String> parameters = getCertificateSubjectParameters(parameters);
        return parameters.get(CN);
    }
```
This would reduce the duplication even further. :)